### PR TITLE
[App] Introduce Multi Node Component

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds a Multi-Node component that is agnostic of the framework

```python
import lightning as L
from lightning.app.components import MultiNode


class AnyDistributedComponent(L.LightningWork):
    def run(
        self,
        master_address: str,
        master_port: int,
        node_rank: int,
    ):
        print(f"ADD YOUR DISTRIBUTED CODE: {master_address} {master_port} {node_rank}")


compute = L.CloudCompute("gpu")
app = L.LightningApp(
    MultiNode(
        AnyDistributedComponent,
        nodes=2,
        cloud_compute=compute,
    )
)
```

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/components/multi_node.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            # 3. Run the user code in a distributed way !
+            self.ws[node_rank].run(
+                master_address=self.ws[0].internal_ip,
+                master_port=self.ws[0].port,
+                node_rank=node_rank,
+            )
+
+            # 4. Stop the machine when finished.
+            if self.ws[node_rank].has_succeeded:
+                self.ws[node_rank].stop()
```
Comment (rohitgr7) [inline src/lightning_app/components/multi_node.py:92]: do you think we should check that all the machines has succeeded and then stop since this is mult-node?

Comment (rohitgr7) [reply src/lightning_app/components/multi_node.py:92]: or if any of them failed, then stop all?

Comment (tchaton) [reply src/lightning_app/components/multi_node.py:93]: Better to stop them at soon as finished. Cost is important.

**Review by lantiga — APPROVED** [review]

File: tests/tests_app_examples/test_multi_node.py [hunk 2]
```diff
@@ -27,3 +27,27 @@ def test_multi_node_example():
     result = application_testing(LightningTestMultiNodeApp, command_line)
```
Comment (awaelchli) [inline tests/tests_app_examples/test_multi_node.py:27]: Can we add a test on the number of works that get created/started based on the value of the `num_nodes` argument?

File: src/lightning_app/core/work.py [hunk 3]
```diff
@@ -272,7 +272,7 @@ def statuses(self) -> List[WorkStatus]:
     @property
     def has_started(self) -> bool:
         """Return whether the work has started."""
-        return self.status.stage != WorkStageStatus.NOT_STARTED
+        return self.status.stage == WorkStageStatus.STARTED
```
Comment (awaelchli) [inline src/lightning_app/core/work.py:275]: Can you explain this change? I don't think it is obvious. In the past, we struggled defining the negation of started and not started etc.

File: src/lightning_app/components/multi_node.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            import torch
+
+            import lightning as L
+            from lightning.components import MultiNode
+
+            class AnyDistributedComponent(L.LightningWork):
+                def run(
+                    self,
+                    master_address: str,
+                    master_port: int,
```
Comment (awaelchli) [inline src/lightning_app/components/multi_node.py:31]: In Lightning Trainer, we have moved away from the word "master'. We use main_address, main_port.

File: src/lightning_app/components/multi_node.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        if all(not w.has_started for w in self.ws):
+            return
+
+        # Loop over all node machines
+        for node_rank in range(self.nodes):
+
+            # 3. Run the user code in a distributed way !
+            self.ws[node_rank].run(
+                master_address=self.ws[0].internal_ip,
+                master_port=self.ws[0].port,
```
Comment (awaelchli) [inline src/lightning_app/components/multi_node.py:87]: same comment as above

File: src/lightning_app/components/multi_node.py (hunk 4 above)
Comment (tchaton) [reply src/lightning_app/components/multi_node.py:31]: Done !

File: src/lightning_app/components/multi_node.py (hunk 5 above)
Comment (tchaton) [reply src/lightning_app/components/multi_node.py:87]: Done !

File: src/lightning_app/core/work.py (hunk 3 above)
Comment (tchaton) [reply src/lightning_app/core/work.py:275]: Removed.

File: src/lightning_app/components/multi_node.py (hunk 4 above)
Comment (williamFalcon) [reply src/lightning_app/components/multi_node.py:31]: yes! thanks for calling this out @awaelchli

File: tests/tests_app_examples/test_multi_node.py (hunk 2 above)
Comment (tchaton) [reply tests/tests_app_examples/test_multi_node.py:27]: Done !

**Review by awaelchli — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by manskx — APPROVED** [review]
🚀
